substrate/frame/identity/src/lib.rs

```diff
@@ -886,6 +966,180 @@ pub mod pallet {
+			ensure!(
+				UsernameAuthorities::<T>::contains_key(&authority),
+				Error::<T>::NotUsernameAuthority
+			);
+			UsernameAuthorities::<T>::remove(&authority);
+			Self::deposit_event(Event::AuthorityRemoved { authority });
+			Ok(())
+		}
+
+		/// Set the username for `who`. Must be called by a username authority.
+		#[pallet::call_index(17)]
+		#[pallet::weight(T::WeightInfo::set_username_for())]
+		pub fn set_username_for(
+			origin: OriginFor<T>,
+			who: AccountIdLookupOf<T>,
+			username: Vec<u8>,
+			signature: Option<T::OffchainSignature>,
+		) -> DispatchResult {
+			// Ensure origin is a Username Authority and has an allocation. Decrement their
+			// allocation by one.
+			let sender = ensure_signed(origin)?;
+			let suffix = UsernameAuthorities::<T>::try_mutate(
+				&sender,
+				|maybe_authority| -> Result<Suffix, DispatchError> {
+					let properties =
+						maybe_authority.as_mut().ok_or(Error::<T>::NotUsernameAuthority)?;
+					ensure!(properties.allocation > 0, Error::<T>::NoAllocation);
+					properties.allocation.saturating_dec();
+					Ok(properties.suffix.clone())
+				},
+			)?;
+
+			// Ensure that the username only contains allowed characters. We already know the suffix
+			// does.
+			let username_length = username.len().saturating_add(suffix.len()) as u32;
+			Self::validate_username(&username, Some(username_length))?;
+
+			// Concatenate the username with suffix and cast into a BoundedVec. Should be infallible
+			// since we already ensured it is below the max length.
+			let mut full_username =
+				Vec::with_capacity(username.len().saturating_add(suffix.len()).saturating_add(1));
+			full_username.extend(username);
+			full_username.extend(b".");
+			full_username.extend(suffix);
+			let bounded_username =
+				Username::try_from(full_username).map_err(|_| Error::<T>::InvalidUsername)?;
+
+			// Usernames must be unique. Ensure it's not taken.
+			ensure!(
+				!AccountOfUsername::<T>::contains_key(bounded_username.clone()),
+				Error::<T>::UsernameTaken
+			);
+
+			// Insert or queue.
+			let who = T::Lookup::lookup(who)?;
+			if let Some(s) = signature {
+				// Account has pre-signed an authorization. Verify the signature provided and grant
+				// the username directly.
+				let encoded = Encode::encode(&bounded_username.to_vec());
+				Self::validate_signature(&encoded, &s, &who)?;
+				Self::insert_username(&who, bounded_username);
+			} else {
+				// The user must accept the username, therefore, queue it.
+				Self::queue_acceptance(&who, bounded_username);
+			}
+			Ok(())
+		}
+
+		/// Accept a given username that an `authority` granted. The call must include the full
+		/// username, as in `username.suffix`.
+		#[pallet::call_index(18)]
+		#[pallet::weight(T::WeightInfo::accept_username())]
+		pub fn accept_username(
+			origin: OriginFor<T>,
+			username: Username,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let (approved_for, _) =
```

Comment:
Need to check the expiration date?

Reply:
IMO no, the expiration is to make sure storage gets cleaned up. If it's expired, they got lucky, but the end result is the same, that the username is removed from `PendingUsernames`.

Reply:
Yea i was just wondering how it was meant. It could also be meant as insurance to the authority that their grant is only valid for a limited time, since they cannot retract it.

Reply:
Ah I see. Yeah in that case we're relying on people to actively patrol expired names and remove them so they don't linger forever.